shelley-ma/formal-spec/value.tex

```diff
@@ -1,77 +1,239 @@
+These types are convertible to and from the $\Value$ type, and appear in other
+parts of the system as a multi-asset representation which is more suitable in particular
+usecases.
+
+\subsection{$\ValMonoid$}
+
+Figure \ref{fig:ValMonoid} gives names and types of functions which
+make up the $\ValMonoid$ structure.
+
+%%
+%% Figure ValMonoid and its Functions
+%%
+\begin{figure}[htb]
+  \begin{align*}
+      & \mathsf{ValType} ~\in~ \mathsf{Type}  \\
+      & \text{A type}
+      \nextdef
+      %
+      & \fun{zero} ~\in~ \mathsf{ValType} \\
+      & \text{Additive identity of $\mathsf{ValType}$}
+      \nextdef
+      %
+      & (+) ~\in~ \mathsf{ValType} \to \mathsf{ValType} \to \mathsf{ValType}\\
+      & \text{Addition on $\mathsf{ValType}$}
+      \nextdef
+      %
+      & (*) ~\in~ \mathsf{Integer} \to \mathsf{ValType} \to \mathsf{ValType}\\
+      & \text{Scale $\mathsf{ValType}$ value by integral constant}
+      \nextdef
+      %
+      & (-) ~\in~ \mathsf{ValType} \to \mathsf{ValType} \to \mathsf{ValType}\\
+      & v_1~-~v_2 ~=~ v_1~+~((-1)~*~v_2)\\
+      & \text{Subtraction on $\mathsf{ValType}$}
+      \nextdef
+      %
+      & \fun{isZero} ~\in~ \mathsf{ValType} \to \Bool\\
+      & \fun{isZero}~v~=~(v~==~\fun{zero}) \\
+      & \text{Is a given $\mathsf{ValType}$ term the additive identity?}
+      \nextdef
+      %
+      & \fun{coin} ~\in~ \mathsf{ValType} \to \Coin\\
+      & \text{Return the Ada contained in a $\mathsf{ValType}$ term}
+      \nextdef
+      %
+      & \fun{inject} ~\in~ \Coin \to \mathsf{ValType}\\
+      & \text{Create a $\mathsf{ValType}$ term containing only this amount of Ada}
+      \nextdef
+      %
+      & \fun{adaOnly} ~\in~ \mathsf{ValType} \to \Bool \\
+      & \fun{adaOnly} =
+        \begin{cases}
+          \True & \fun{inject} \circ \fun{coin}~ = ~\fun{id}_{\mathsf{ValType}} \\
+          \False & \text{otherwise}
+        \end{cases} \\
+      & \text{Check if a given $\mathsf{ValType}$ term contains only Ada}
+      \nextdef
+      %
+      & \fun{modifyCoin} ~\in~ (\Coin \to \Coin) \to (\mathsf{ValType} \to \mathsf{ValType})\\
+      & \text{Modify the Ada part of $\mathsf{ValType}$}
+      \nextdef
+      %
+      & \fun{size} ~\in~ \mathsf{ValType} \to \MemoryEstimate\\
+      & \text{Return the size in words of a $\mathsf{ValType}$ term}
+      \nextdef
+      %
+      & \fun{pointwise} ~\in~ (\Integer \to \Integer \to \Bool) \to (\mathsf{ValType} \to \mathsf{ValType} \to \Bool)\\
+      & \text{Pointwise comparison of two $\mathsf{ValType}$ terms}
+  \end{align*}
+  \caption{The $\ValMonoid$ Structure}
+  \label{fig:ValMonoid}
+\end{figure}
+
+Note that any $\ValMonoid$ must also have the following function, which gives
+the size of the serialized representation of a term of the underlying type.
+This function is used in the UTXO rule.
+We do not include it in the monoid definition because it cannot be defined
+explicity separately from concrete implementation.
```

Comment:
This should actually be part of the definition of `ValMonoid`. If we were to define it, it would have different implementations for each type

Reply:
I think it should not because
- we cant/wont actually define this function in the spec because it is a Haskell function that is opaque to the ledger
- even if we dig up its specification and give it here, it is misleading to pretend that this spec of `serSize` is defined as part of specifying a `ValMonoid` rather than exists for all serializable types

Reply:
changed it to the computation of a serialized representation